docs/docs/authentication-tutorial.md

```diff
@@ -2,7 +2,445 @@
+        >
+          <html lang="en" />
+        </Helmet>
+        <Header siteTitle={data.site.siteMetadata.title} />
+        <div
+          style={{
+            margin: "0 auto",
+            maxWidth: 960,
+            padding: "0px 1.0875rem 1.45rem",
+            paddingTop: 0,
+          }}
+        >
+          <NavBar />
+          {children}
+        </div>
+      </>
+    )}
+  />
+)
+
+Layout.propTypes = {
+  children: PropTypes.node.isRequired,
+}
+
+export default Layout
+```
+
+Lastly, change the index page to include this new content:
+
+```jsx{9-11}:title=src/pages/index.js
+import React from "react"
+import { Link } from "gatsby"
+
+import Layout from "../components/layout"
+
+const IndexPage = () => (
+  <Layout>
+    <h1>Hi people</h1>
+    <p>
+      You should <Link to="/">log in</Link> to see restricted content
+    </p>
+  </Layout>
+)
+
+export default IndexPage
+```
+
+## Authentication service
+
+For this tutorial you will use a hardcoded user/password. Create the folder `src/services` and add the following content to the file `auth.js`:
+
+```javascript:title=src/services/auth.js
+export const isBrowser = () => typeof window !== "undefined"
+
+export const getUser = () =>
+  isBrowser() && window.localStorage.gatsbyUser
+    ? JSON.parse(window.localStorage.gatsbyUser)
+    : {}
+
+const setUser = user => (window.localStorage.gatsbyUser = JSON.stringify(user))
+
+export const handleLogin = ({ username, password }) => {
+  if (username === `john` && password === `pass`) {
+    return setUser({
+      username: `john`,
+      name: `Johnny`,
+      email: `johnny@example.org`,
+    })
+  }
+
+  return false
+}
+
+export const isLoggedIn = () => {
+  const user = getUser()
+
+  return !!user.username
+}
+
+export const logout = callback => {
+  setUser({})
+  callback()
+}
+```
+
+## Creating client-only routes
+
+At the beginning of this tutorial, you created a "default" Gatsby site, which includes the `@reach/router` library. Now, using the [@reach/router](https://reach.tech/router/) library, you can create routes available only to logged-in users. This library is used by Gatsby under the hood, so you don't even have to install it.
+
+First, edit `gatsby-node.js`. You will define that any route that starts with `/app/` is part of your restricted content and the page will be created on demand:
+
+```javascript:title=gatsby-config.js
+// Implement the Gatsby API “onCreatePage”. This is
+// called after every page is created.
+exports.onCreatePage = async ({ page, actions }) => {
+  const { createPage } = actions
+
+  // page.matchPath is a special key that's used for matching pages
+  // only on the client.
+  if (page.path.match(/^\/app/)) {
+    page.matchPath = "/app/*"
+
+    // Update the page.
+    createPage(page)
+  }
+}
+```
+
+> Note: There is a convenient plugin that already does this work for you: [gatsby-plugin-create-client-paths](https://www.gatsbyjs.org/packages/gatsby-plugin-create-client-paths/)
```

Comment:
I didn't even know this 😬 Neat, thanks for including this!

Reply:
:+1: